### PR TITLE
Add support for boost::multi_array of EigenMatrix #261

### PR DESCRIPTION
**Pull request by alexsavulescu**

### CODE REVIEW

File: include/highfive/bits/H5Converter_misc.hpp [hunk 1]
```diff
@@ -388,7 +402,59 @@ struct data_converter<std::vector<Eigen::Matrix<T,M,N>>, void> {
...
+                const auto VEC_M = vec.origin()->rows(), VEC_N = vec.origin()->cols();
+                for (auto v = vec.origin(); v < vec.origin() + vec.num_elements(); ++v) {
+                    assert(v->rows() == VEC_M && v->cols() == VEC_N);
+                    *v = Eigen::Map<Eigen::Matrix<T, M, N>>(start, VEC_M, VEC_N);
+                    start += VEC_M * VEC_N;
+                }
+            } else {
+                std::ostringstream ss;
+                ss << "Dynamic size(-1) used without pre-defined multi_array data layout.\n"
+                   << "Initiliaze vector elements using  MatrixXd::Zero";
```
Comment (tristan0x) [inline include/highfive/bits/H5Converter_misc.hpp]:
```suggestion
                   << "Initialize vector elements using  MatrixXd::Zero";
```

File: include/highfive/bits/H5Converter_misc.hpp [hunk 2]
```diff
@@ -388,7 +402,59 @@ struct data_converter<std::vector<Eigen::Matrix<T,M,N>>, void> {
...
+            if (vec.origin()->rows() > 0 && vec.origin()->cols() > 0) {
+                const auto VEC_M = vec.origin()->rows(), VEC_N = vec.origin()->cols();
+                for (auto v = vec.origin(); v < vec.origin() + vec.num_elements(); ++v) {
+                    assert(v->rows() == VEC_M && v->cols() == VEC_N);
+                    *v = Eigen::Map<Eigen::Matrix<T, M, N>>(start, VEC_M, VEC_N);
+                    start += VEC_M * VEC_N;
+                }
+            } else {
+                std::ostringstream ss;
+                ss << "Dynamic size(-1) used without pre-defined multi_array data layout.\n"
```
Comment (tristan0x) [inline include/highfive/bits/H5Converter_misc.hpp]: The `ostringstream` is not required. You can simply use the preprocessor to concatenate strings.

throw DataSetException("error message line 1\n"
                                        "error message line 2");

File: include/highfive/bits/H5Dataspace_misc.hpp [hunk 3]
```diff
@@ -206,6 +206,18 @@ DataSpace::From(const std::vector<Eigen::Matrix<Value, M, N>>& vec) {
     return DataSpace(dims);
 }
 
+#ifdef H5_USE_BOOST
+template <typename Value, int M, int N, size_t Dims>
+inline DataSpace DataSpace::From(const boost::multi_array<Eigen::Matrix<Value, M, N>, Dims>& vec) {
+    std::vector<size_t> dims(Dims);
```
Comment (tristan0x) [inline include/highfive/bits/H5Dataspace_misc.hpp]: Using this `vector` constructor prevents the `for` loop
```
std::vector<size_t> dims(vec.shape(), vec.shape() + Dims)
```

File: include/highfive/bits/H5Converter_misc.hpp [hunk 4]
```diff
@@ -388,7 +402,59 @@ struct data_converter<std::vector<Eigen::Matrix<T,M,N>>, void> {
...
+struct data_converter<boost::multi_array<Eigen::Matrix<T, M, N>, Dims>, void> {
+    typedef typename boost::multi_array<Eigen::Matrix<T, M, N>, Dims> MultiArrayEigen;
+
+    inline data_converter(MultiArrayEigen&, DataSpace& space)
+        : _dims(space.getDimensions())
+        , _space(space) {
+        assert(_dims.size() == Dims);
+    }
+
+    inline typename type_of_array<T>::type* transform_read(MultiArrayEigen& /*array*/) {
```
Comment (tristan0x) [inline include/highfive/bits/H5Converter_misc.hpp]:
```suggestion
    inline typename type_of_array<T>::type* transform_read(const MultiArrayEigen& /*array*/) {
```

File: include/highfive/bits/H5Converter_misc.hpp [hunk 5]
```diff
@@ -388,7 +402,59 @@ struct data_converter<std::vector<Eigen::Matrix<T,M,N>>, void> {
     DataSpace& _space;
 };
 
+#ifdef H5_USE_BOOST
+template <typename T, int M, int N, std::size_t Dims>
+struct data_converter<boost::multi_array<Eigen::Matrix<T, M, N>, Dims>, void> {
+    typedef typename boost::multi_array<Eigen::Matrix<T, M, N>, Dims> MultiArrayEigen;
+
+    inline data_converter(MultiArrayEigen&, DataSpace& space)
```
Comment (tristan0x) [inline include/highfive/bits/H5Converter_misc.hpp]:
```suggestion
    inline data_converter(const MultiArrayEigen&, DataSpace& space)
```

File: include/highfive/bits/H5Dataspace_misc.hpp (hunk 3 above)
Comment (alexsavulescu) [reply include/highfive/bits/H5Dataspace_misc.hpp]: Indeed, also updated another part in the code with this.

**Pull request merged**